core/src/main/java/kafka/server/share/SharePartitionManager.java

```diff
@@ -533,178 +538,52 @@ void maybeProcessFetchQueue() {
-                // is in-flight and should release the lock. But it's safe to release the lock as
-                // the lock on share partition already exists which facilitates correct behaviour
-                // with multiple requests from queue being processed.
-                releaseProcessFetchQueueLock();
-                if (!fetchQueue.isEmpty())
-                    maybeProcessFetchQueue();
-                return;
-            }
+            Set<Object> delayedShareFetchWatchKeys = new HashSet<>();
+            delayedShareFetchWatchKeys.add(new DelayedShareFetchKey(
+                shareFetchPartitionData.partitionMaxBytes.keySet(),
+                shareFetchPartitionData.groupId,
+                shareFetchPartitionData.memberId));
 
-            log.trace("Fetchable share partitions data: {} with groupId: {} fetch params: {}",
-                topicPartitionData, shareFetchPartitionData.groupId, shareFetchPartitionData.fetchParams);
-
-            replicaManager.fetchMessages(
-                shareFetchPartitionData.fetchParams,
-                CollectionConverters.asScala(
-                    topicPartitionData.entrySet().stream().map(entry ->
-                        new Tuple2<>(entry.getKey(), entry.getValue())).collect(Collectors.toList())
-                ),
-                QuotaFactory.UnboundedQuota$.MODULE$,
-                responsePartitionData -> {
-                    log.trace("Data successfully retrieved by replica manager: {}", responsePartitionData);
-                    List<Tuple2<TopicIdPartition, FetchPartitionData>> responseData = CollectionConverters.asJava(
-                        responsePartitionData);
-                    processFetchResponse(shareFetchPartitionData, responseData).whenComplete(
-                        (result, throwable) -> {
-                            if (throwable != null) {
-                                log.error("Error processing fetch response for share partitions", throwable);
-                                shareFetchPartitionData.future.completeExceptionally(throwable);
-                            } else {
-                                shareFetchPartitionData.future.complete(result);
-                            }
-                            // Releasing the lock to move ahead with the next request in queue.
-                            releaseFetchQueueAndPartitionsLock(shareFetchPartitionData.groupId, topicPartitionData.keySet());
-                        });
-                    return BoxedUnit.UNIT;
-                });
+            // Add the share fetch to the delayed share fetch purgatory to process the fetch request.
+            addDelayedShareFetch(new DelayedShareFetch(shareFetchPartitionData, replicaManager, partitionCacheMap),
+                delayedShareFetchWatchKeys);
 
+            // Release the lock so that other threads can process the queue.
+            releaseProcessFetchQueueLock();
```

Comment:
If `maybeProcessFetchQueue` throws an exception, you might find the catch block releases the fetch queue lock which has been obtained by another thread.

Reply:
Because we are just using an AtomicBoolean for the lock, we do need to be squeaky clean on managing the lock status because there's no concept of ownership from the point of view of the JVM.

Reply:
Hi @AndrewJSchofield , I have reverted the changes here since it was incorrect, so the logic is same as earlier.